tests/db_tests/test_constraints.py

```diff
@@ -119,6 +121,20 @@ def test_output_constraint_project_xor_programme(seeded_test_client_rollback):
         db.session.commit()
 
 
+def test_project_geospatial_association(seeded_test_client_rollback):
```

Comment:
I think this is the wrong fixture? This one intended for non-committed changes, but in this case you do commit, so it's in danger of leaking into other test scopes.

Reply:
Hm, this sounds inherently risky. Each test case should be fully isolated and rolled back by default. I don't think any test case should be able to commit data that lasts beyond the test run - unless we very explicitly expect the test database to be in a messy state and tests are built specifically to handle that.

Reply:
`test_client_reset` says in the docstring that it's intended for tests that need an empty DB and need to commit DB changes as part of their execution, and as it's function scoped it then resets all the DB tables to empty. I copied that setup for this new fixture and just added the geospatial data seeding at the top, so it should reset the db when the test ends.

`test_client_reset` is the one that many of the other integration tests that commit data to the DB used and needed to be updated to this fixture so they can make use of the seeded geospatial table, but as it's function scoped and resets any changes after the test there shouldn't be any leaking? Or have I misunderstood the fixtures (which is entirely possible!)?

Reply:
Sam, I think we're on the same page here, I maybe just didn't explain clearly:
- We have test fixtures that completely reset the DB at a function level, to be used if any DB changes are committed in the test.
- We also have test fixtures that rollback uncommitted changes in a test - not intended to be used anywhere where commit is used. This is to prevent the complete reset of DB for every test where it's not always necassary.

In this case, my comment was saying that I think Marc used a test fixture that only rolls back uncommitted changes, but the test does a commit, the fixture #I could see in use was `seeded_test_client_rollback` but should have bee none at function scope for resetting the DB.

Reply:
Hey both, yep you're both right, I've used the wrong fixture here - will update it to use reset instead of rollback. Thanks for catching!

Reply:
I'm a bit torn. It feels like a chunk of mental overhead to have to work out whether the code under test is calling a `commit` at any point, and unless we have a very clear definition of when commits happen, feels again like something that could be refactored in the future and then change the semantics of the rest without anyone really spotting it.

Generally I'd probably want every test, by default, to rollback any changes that were made during the test - 'committed' or not.

If speed is a concern we can easily start looking at parallelising them (and always better to do that sooner rather than later so that you catch structural issues with concurrent test runes early).